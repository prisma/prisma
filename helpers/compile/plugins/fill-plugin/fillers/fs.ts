export function existsSync() {
  return false
}

/**
 * A stub for fs for tryLoadEnv not to attempt loading .env
 */
const fs = {
  existsSync() {
    return false
  },
}

export default fs
