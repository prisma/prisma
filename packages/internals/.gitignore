node_modules
query-engine*
libquery_engine*
introspection-engine*
migration-engine*
prisma-fmt*
libquery*
/engines
/debug.d.ts
/debug.js
/debug.js.map
pnpm-lock.yaml
yarn-error.log
.DS_Store

