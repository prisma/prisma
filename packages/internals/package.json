{
  "name": "@prisma/internals",
  "version": "0.0.0",
  "description": "This package is intended for Prisma's internal use",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/prisma/prisma.git",
    "directory": "packages/internals"
  },
  "homepage": "https://www.prisma.io",
  "author": "Tim Suchanek <suchanek@prisma.io>",
  "bugs": "https://github.com/prisma/prisma/issues",
  "license": "Apache-2.0",
  "scripts": {
    "dev": "DEV=true node -r esbuild-register helpers/build.ts",
    "build": "node -r esbuild-register helpers/build.ts",
    "test": "jest --verbose",
    "prepublishOnly": "pnpm run build"
  },
  "files": [
    "README.md",
    "dist",
    "!**/libquery_engine*",
    "!dist/get-generators/engines/*",
    "scripts"
  ],
  "devDependencies": {
    "@swc/core": "1.2.204",
    "@swc/jest": "0.2.24",
    "@types/jest": "29.2.4",
    "@types/node": "12.20.55",
    "@types/resolve": "1.20.2",
    "esbuild": "0.15.13",
    "jest": "29.3.1",
    "jest-junit": "15.0.0",
    "mock-stdin": "1.0.0",
    "ts-node": "10.9.1",
    "typescript": "4.8.4",
    "yarn": "1.22.19"
  },
  "dependencies": {
    "@prisma/debug": "workspace:*",
    "@prisma/engine-core": "workspace:*",
    "@prisma/engines": "workspace:*",
    "@prisma/fetch-engine": "workspace:*",
    "@prisma/generator-helper": "workspace:*",
    "@prisma/get-platform": "workspace:*",
    "@prisma/prisma-fmt-wasm": "4.9.0-39.df61b5f5ab3f27359e273f1cba4e5a9c7cb8485a",
    "archiver": "5.3.1",
    "arg": "5.0.2",
    "chalk": "4.1.2",
    "checkpoint-client": "1.1.21",
    "cli-truncate": "2.1.0",
    "dotenv": "16.0.3",
    "escape-string-regexp": "4.0.0",
    "execa": "5.1.1",
    "find-up": "5.0.0",
    "fp-ts": "2.13.1",
    "fs-extra": "11.1.0",
    "fs-jetpack": "5.1.0",
    "global-dirs": "3.0.0",
    "globby": "11.1.0",
    "has-yarn": "2.1.0",
    "is-windows": "^1.0.2",
    "is-wsl": "^2.2.0",
    "new-github-issue-url": "0.2.1",
    "node-fetch": "2.6.7",
    "open": "7",
    "ora": "5.4.1",
    "p-map": "4.0.0",
    "prompts": "2.4.2",
    "read-pkg-up": "7.0.1",
    "replace-string": "3.1.0",
    "resolve": "1.22.1",
    "string-width": "4.2.3",
    "strip-ansi": "6.0.1",
    "strip-indent": "3.0.0",
    "temp-dir": "2.0.0",
    "temp-write": "4.0.0",
    "tempy": "1.0.1",
    "terminal-link": "2.1.1",
    "tmp": "0.2.1",
    "ts-pattern": "^4.0.1"
  },
  "sideEffects": false
}
