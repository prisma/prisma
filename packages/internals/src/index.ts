export { checkUnsupportedDataProxy } from './cli/checkUnsupportedDataProxy'
export { getGeneratorSuccessMessage } from './cli/getGeneratorSuccessMessage'
export {
  getPrismaConfigFromPackageJson,
  getRelativeSchemaPath,
  getSchema,
  getSchemaDir,
  getSchemaDirSync,
  getSchemaPath,
  getSchemaPathFromPackageJson,
  getSchemaPathFromPackageJsonSync,
  getSchemaPathSync,
  getSchemaSync,
} from './cli/getSchema'
export { getCLIPathHash, getProjectHash } from './cli/hashes'
export { unknownCommand } from './cli/Help'
export { HelpError } from './cli/Help'
export type {
  Command,
  Commands,
  CompiledGeneratorDefinition,
  Dictionary,
  GeneratorConfig,
  GeneratorDefinition,
  GeneratorDefinitionWithPackage,
  GeneratorFunction,
  GeneratorOptions,
} from './cli/types'
export { arg, format, isError } from './cli/utils'
export { ClientEngineType, DEFAULT_CLIENT_ENGINE_TYPE, getClientEngineType } from './client/getClientEngineType'
export { credentialsToUri, protocolToConnectorType, uriToCredentials } from './convertCredentials'
export * from './engine-commands'
export { Generator } from './Generator'
export type { ProviderAliases } from './get-generators/getGenerators'
export { getGenerator, getGenerators } from './get-generators/getGenerators'
export { getPackedPackage } from './getPackedPackage'
export { highlightDatamodel, highlightSql, highlightTS } from './highlight/highlight'
export type { IntrospectionSchemaVersion, IntrospectionWarnings } from './IntrospectionEngine'
export { IntrospectionEngine } from './IntrospectionEngine'
export * as logger from './logger'
export type { MigrateEngineLogLine } from './migrateEngineCommands'
export { canConnectToDatabase, createDatabase, dropDatabase } from './migrateEngineCommands'
export { MigrateEngineExitCode } from './migrateEngineCommands'
export { ErrorArea, isRustPanic, RustPanic } from './panic'
export type { GeneratorPaths } from './predefinedGeneratorResolvers'
export { BinaryType } from './resolveBinary'
export { engineEnvVarMap, resolveBinary } from './resolveBinary'
export { sendPanic } from './sendPanic'
export type { DatabaseCredentials } from './types'
export { assertNever } from './utils/assertNever'
export { callOnce } from './utils/callOnce'
export { canPrompt } from './utils/canPrompt'
export { drawBox } from './utils/drawBox'
export { extractPreviewFeatures } from './utils/extractPreviewFeatures'
export { formatms } from './utils/formatms'
export { formatTable } from './utils/formatTable'
export { getCommandWithExecutor } from './utils/getCommandWithExecutor'
export type { EnvPaths } from './utils/getEnvPaths'
export { getEnvPaths } from './utils/getEnvPaths'
export { version } from './utils/getVersionFromPackageJson'
export { handlePanic } from './utils/handlePanic'
export { hasOwnProperty } from './utils/hasOwnProperty'
export { isCi } from './utils/isCi'
export { isCurrentBinInstalledGlobally } from './utils/isCurrentBinInstalledGlobally'
export { isInteractive } from './utils/isInteractive'
export { isPromiseLike } from './utils/isPromiseLike'
export { jestConsoleContext, jestContext, jestProcessContext } from './utils/jestContext'
export { keyBy } from './utils/keyBy'
export { link } from './utils/link'
export { load } from './utils/load'
export { loadEnvFile } from './utils/loadEnvFile'
export { mapObjectValues } from './utils/mapObjectValues'
export { mapPreviewFeatures } from './utils/mapPreviewFeatures'
export { maskSchema } from './utils/maskSchema'
export { missingGeneratorMessage } from './utils/missingGeneratorMessage'
export { parseBinaryTargetsEnvValue, parseEnvValue } from './utils/parseEnvValue'
export { pick } from './utils/pick'
export { platformRegex } from './utils/platformRegex'
export { printConfigWarnings } from './utils/printConfigWarnings'
export { serializeQueryEngineName } from './utils/serializeQueryEngineName'
export { createSpinner } from './utils/spinner'
export type { Position } from './utils/trimBlocksFromSchema'
export { trimBlocksFromSchema, trimNewLine } from './utils/trimBlocksFromSchema'
export { tryLoadEnvs } from './utils/tryLoadEnvs'
export { warnOnce } from './warnOnce'
export * as wasm from './wasm'
export type { Platform } from '@prisma/get-platform'
export { getNodeAPIName, getPlatform } from '@prisma/get-platform'
