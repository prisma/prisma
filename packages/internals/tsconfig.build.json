{
  "extends": "../../tsconfig.build.regular.json",
  "compilerOptions": {
    "outDir": "dist",
    "typeRoots": ["./node_modules/@types", "./typings"]
  },
  "include": ["src"]
}
