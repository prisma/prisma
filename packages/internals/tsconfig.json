{
  "extends": "../../tsconfig.json",
  "include": ["./**/*"],
  "exclude": ["./**/__fixtures__/**/*"]
}
