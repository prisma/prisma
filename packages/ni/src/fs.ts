import { resolve } from 'node:path'
import fs from 'node:fs'
import type { RunnerContext } from './runner'

export function getPackageJSON(ctx?: RunnerContext): any {
  const cwd = ctx?.cwd ?? process.cwd()
  const path = resolve(cwd, 'package.json')

  if (fs.existsSync(path)) {
    try {
      const raw = fs.readFileSync(path, 'utf-8')
      const data = JSON.parse(raw)
      return data
    }
    catch (e) {
      if (!ctx?.programmatic) {
        console.warn('Failed to parse package.json')
        process.exit(1)
      }

      throw e
    }
  }
}
