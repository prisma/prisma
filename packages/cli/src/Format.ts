import { arg, Command, format, formatms, formatSchema, getDMMF, HelpError } from '@prisma/internals'
import { getSchemaPathAndPrint } from '@prisma/migrate'
import chalk from 'chalk'
import fs from 'fs'

/**
 * $ prisma format
 */
export class Format implements Command {
  public static new(): Format {
    return new Format()
  }

  private static help = format(`
Format a Prisma schema.

${chalk.bold('Usage')}

  ${chalk.dim('$')} prisma format [options]

${chalk.bold('Options')}

  -h, --help   Display this help message
    --schema   Custom path to your Prisma schema

${chalk.bold('Examples')}

With an existing Prisma schema
  ${chalk.dim('$')} prisma format

Or specify a Prisma schema path
  ${chalk.dim('$')} prisma format --schema=./schema.prisma

  `)

  public async parse(argv: string[]): Promise<string | Error> {
    const before = Date.now()
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--schema': String,
      '--telemetry-information': String,
    })

    if (args instanceof Error) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    const schemaPath = await getSchemaPathAndPrint(args['--schema'])

    const output = await formatSchema({ schemaPath })

    try {
      // Validate whether the formatted output is a valid schema
      await getDMMF({
        datamodel: output,
      })
    } catch (e) {
      console.error('') // empty line for better readability
      throw e
    }

    fs.writeFileSync(schemaPath, output)
    const after = Date.now()

    return `Formatted ${chalk.underline(schemaPath)} in ${formatms(after - before)} ðŸš€`
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${Format.help}`)
    }
    return Format.help
  }
}
