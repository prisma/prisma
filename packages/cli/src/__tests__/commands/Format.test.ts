import { jestConsoleContext, jestContext } from '@prisma/internals'
import fs from 'fs-jetpack'

import { Format } from '../../Format'

const ctx = jestContext.new().add(jestConsoleContext()).assemble()

const originalEnv = { ...process.env }

describe('format', () => {
  beforeEach(() => {
    process.env = { ...originalEnv }
  })
  afterAll(() => {
    process.env = { ...originalEnv }
  })

  it('should add a trailing EOL', async () => {
    ctx.fixture('example-project/prisma')
    await Format.new().parse([])
    expect(fs.read('schema.prisma')).toMatchSnapshot()
  })

  it('should add missing backrelation', async () => {
    ctx.fixture('example-project/prisma')
    await Format.new().parse(['--schema=missing-backrelation.prisma'])
    expect(fs.read('missing-backrelation.prisma')).toMatchSnapshot()
  })

  it('should throw if schema is broken', async () => {
    ctx.fixture('example-project/prisma')
    await expect(Format.new().parse(['--schema=broken.prisma'])).rejects.toThrow()
  })

  it('should succeed and show a warning on stderr (preview feature deprecated)', async () => {
    ctx.fixture('lint-warnings')
    await expect(Format.new().parse(['--schema=preview-feature-deprecated.prisma'])).resolves.toBeTruthy()

    // stderr
    expect(ctx.mocked['console.warn'].mock.calls.join('\n')).toMatchInlineSnapshot(`

            Prisma schema warning:
            - Preview feature "nativeTypes" is deprecated. The functionality can be used without specifying it as a preview feature.
      `)
    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
  })

  it('should throw with an error and show a warning on stderr (preview feature deprecated)', async () => {
    ctx.fixture('lint-warnings')
    await expect(Format.new().parse(['--schema=preview-feature-deprecated-and-error.prisma'])).rejects.toThrow('P1012')

    // stderr
    expect(ctx.mocked['console.warn'].mock.calls.join('\n')).toMatchInlineSnapshot(`

            Prisma schema warning:
            - Preview feature "nativeTypes" is deprecated. The functionality can be used without specifying it as a preview feature.
      `)
    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
  })

  it('should succeed and NOT show a warning when PRISMA_DISABLE_WARNINGS is truthy', async () => {
    ctx.fixture('lint-warnings')

    process.env.PRISMA_DISABLE_WARNINGS = 'true'

    await expect(Format.new().parse(['--schema=preview-feature-deprecated.prisma'])).resolves.toBeTruthy()

    // stderr
    expect(ctx.mocked['console.warn'].mock.calls.join('\n')).toEqual('')
    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toEqual('')
  })
})
