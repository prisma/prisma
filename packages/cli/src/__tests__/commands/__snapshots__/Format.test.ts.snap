// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format should add a trailing EOL 1`] = `
generator client {
  provider = "prisma-client-js"
  output   = "../generated/client"
}

datasource db {
  provider = "sqlite"
  url      = "file:dev.db"
}

model Post {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  title     String
  content   String?
  published Boolean  @default(false)
  author    User     @relation(fields: [authorId], references: [id])
  authorId  Int
}

model Profile {
  id     Int     @id @default(autoincrement())
  bio    String?
  user   User    @relation(fields: [userId], references: [id])
  userId Int     @unique
}

model User {
  id      Int      @id @default(autoincrement())
  email   String   @unique
  name    String?
  posts   Post[]
  profile Profile?
}

`;

exports[`format should add missing backrelation 1`] = `
generator client {
  provider = "prisma-client-js"
  output   = "../generated/client"
}

datasource db {
  provider = "sqlite"
  url      = "file:dev.db"
}

model Post {
  id     Int  @id
  user   User @relation(fields: [userId], references: [id])
  userId Int
}

model User {
  id    Int    @id
  posts Post[]
}

`;
